imports/lib/models/Model.ts

```diff
@@ -0,0 +1,187 @@
+import { Mongo } from 'meteor/mongo';
+import type { Document, IndexSpecification, CreateIndexesOptions } from 'mongodb';
+import { z } from 'zod';
+import { IsInsert, IsUpdate, IsUpsert } from '../schemas/customTypes';
+
+type Selector<T extends Document> = Mongo.Selector<T> | string | Mongo.ObjectID;
+type SelectorToResultType<T extends Document, S extends Selector<T>> =
+  S extends string ? T & { _id: S } :
+  S extends Mongo.ObjectID ? T & { _id: S } :
+  T & { [K in keyof S & keyof T]: S[K] extends T[K] ? S[K] : never };
+
+// This is a re-implementation of ZodObject.partial which doesn't wrap fields
+// which already accept undefined as input
+function optionalPreservingPartial<
+  T extends z.ZodRawShape,
+  UnknownKeys extends z.UnknownKeysParam,
+  Catchall extends z.ZodTypeAny,
+>(schema: z.ZodObject<T, UnknownKeys, Catchall>): z.ZodObject<{
+  [K in keyof T]: undefined extends z.input<T[K]> ? T[K] : z.ZodOptional<T[K]>;
+}, UnknownKeys, Catchall> {
+  const newShape: any = {};
+  for (const [key, fieldSchema] of Object.entries(schema.shape)) {
+    newShape[key] = fieldSchema.isOptional() ? fieldSchema : fieldSchema.optional();
+  }
+
+  return new z.ZodObject({
+    ...schema._def,
+    shape: () => newShape,
+  });
+}
+
+const modifierIsWholeDoc = <T extends Document>(modifier: Mongo.Modifier<T>): modifier is T => {
+  return !Object.keys(modifier).some((k) => k.startsWith('$'));
+};
+
+export const AllModels = new Set<Model<any>>();
+
+class Model<T extends z.ZodObject<any, 'strict'>> {
+  name: string;
+
+  schema: T;
+
+  collection: Mongo.Collection<z.output<T>>;
+
+  constructor(name: string, schema: T) {
+    this.name = name;
+    this.schema = schema;
+    this.collection = new Mongo.Collection(name);
+    AllModels.add(this);
+  }
+
+  async insertAsync(doc: z.input<T>): Promise<z.output<T>['_id']> {
+    const cleaned: z.output<T> = await IsInsert.withValue(true, async () => {
+      return this.schema.parseAsync(doc);
+    });
+    return this.collection.insert(cleaned as any) as any;
+  }
+
+  // As a note, we don't accept full documents as modifiers because our
+  // transform functions don't make sense (e.g. we won't accept a `createdAt`
+  // timestamp, but we expect one to be there)
+  async updateAsync(
+    selector: Selector<z.output<T>>,
+    modifier: Exclude<Mongo.Modifier<z.input<T>>, z.input<T>>,
+    options: {
+      multi?: boolean | undefined;
+      upsert?: boolean | undefined;
+      arrayFilters?: { [identifier: string]: any }[] | undefined;
+    } = {},
+  ): Promise<number> {
+    // Types should prevent passing a full document as a modifier, but verify at
+    // runtime to be sure
+    const rawModifier: Mongo.Modifier<z.input<T>> = modifier as any;
+    if (modifierIsWholeDoc(rawModifier)) {
```

Comment:
I had not realized that we never use `update` with a full document, though it makes sense in hindsight.  I went through all our updateAsync calls to be sure, and yep, every one of them uses `$set` or `$addToSet` or some other operator.  Neat!